Add python bunzip support with cl bunzip as fallback

johnwparent opened the pull request:
Refactor compression strategies and conventions for Windows support. These changes do not actually depend on anything in the Windows feature branch, so this PR is targeted at develop to expediate Windows integration into develop down the road.

Changes: Update BZ2 and unzip unpacking strategies to support zipfiles and move away from reliance on unix CL tools.
 
Discussed updates: Updates to prevent bz2 compressed tarballs from hanging on Windows, and do not require Windows users to have unix cl utilities available on the path. Refactor decompression function signature. Refactor decompression strategies for extension as required by the Windows refactor.

johnwparent committed:
Add bunzip with cl bunzip as fallback

Fixup compression strategy

Support python v2.6

Style fixups

johnwparent commented:
@BetsyMcPhail PTAL. I'm not sure what's going on w.r.t the codecov test, it's registering a number of changes not made on this branch.

johnwparent committed:
Refactor compression util

Refactor for more contsitent useage of decompression strategies
Refactor wrapper functions for extension

refactor Windows decompressor approach

Add tar python approach

Refactor gzip decompressor

close tarball

Style fixups

Add docstrings

Catch tar remnant failure

johnwparent commented:
@zackgalbreath If you're interested.

johnwparent commented:
@BetsyMcPhail @zackgalbreath @tgamblin PTAL as time permits/as interested.

johnwparent commented:
@tgamblin AFAICT this is ready for a merge. In that case, would you prefer the commits be squashed?

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
     return any(path.endswith(t) for t in ALLOWED_ARCHIVE_TYPES)
 
 
+def _tar(archive_file):
```
docstring indicating that this handles gzip, xz natively would be good, also this should probably be called `_untar`.

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
+        tar.extractall()
+        tar.close()
+        if sys.platform == "win32":
+            os.remove(remnant)
```
is this always present on win32 and can it fail?  Should we ignore failures here?

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
+    try:
+        import tarfile
+        if ext in [".xz", ".txz"]:
+            import lzma  # noqa # novermin
```
Add a comment saying that this is done b/c if the import succeeds, then `tarfile` can handle xz files.

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
+    _, ext = os.path.splitext(archive_file)
+    try:
+        import tarfile
+        if ext in [".xz", ".txz"]:
```
what happens for `.Z` (7zip) files?  I think they fall through this and Python tries and failed to use `tarfile` on them, then doesn't fall back to using `tar`.  Is that right?  Can you test this with `spack stage gblocks` (which I think is the only thing with a 7zip URL right now)?

johnwparent replied:
My apologies, I completely forgot about 7zip, I'll add a catch for that. At the moment, you're right, currently it tries and fails with Python's native tarfile. In that vein, do we also want to add conditional support for the .7z extension now that Windows support is moving along?

tgamblin replied:
I looked into this a bit more and I think I'm wrong about what `.tar.Z` is.  It's actually not 7zip.  I think I was confused because I looked it up and the recommended program to use to decompress it on windows is 7zip.  See these:

* `.tar.Z` files: https://kb.iu.edu/d/acsy 
* `.Z` extension: https://kb.iu.edu/d/abck

The file signature is 1f 9d 90 -- which [according to wikipedia](https://en.wikipedia.org/wiki/List_of_file_signatures) is LZW compressed.

So yes we should have a special case here for `.Z` (and `.z` which I guess is also zipped).  I'm curious now whether they can be handled by either `zipfile` or `gzip` or some other core python module... but I couldn't really find anything.  So I you can't find anything, then yes I would fail this over to system tar (which seems to handle it fine).

On 7zip itself: yeah we should probably support it, but let's not do it in this PR since nothing currently uses it.  I am not sue what packages we would want to have 7zip downloads for, since it doesn't look like there is a common, builtin 7zip decompressor on linux or macOS.

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
+    return b
+
+
+def decompressor_for(path, ext=None):
```
docstring

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
+    return destination_abspath
+
+
+def composer(funcA):
```
docstring

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
         archive_file (str): absolute path of the file to be decompressed
     """
-    import gzip
     decompressed_file = os.path.basename(archive_file.strip('.gz'))
```
what about tgz files?

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +24,126 @@ def allowed_archive(path):
+    Args:
+        archive_file (str): absolute path to the bz2 archive to be decompressed
+    """
+    decompressed_file = os.path.basename(archive_file.strip(".bz2"))
```
what about tbz2 files?

tgamblin reviewed (changes requested):
This looks ok, but it really needs a lot more tests.  I think we need some parameterized tests that construct simple tar.gz, tar.bz2, tar.Z, tbz2, tgz, etc. files and ensure that both methods (python native and tar) are used on them.  The tests can be marked `skipif` if compression tools like bzip2 or xzip are unavailable, and we'd just need to make sure the right stuff is available in the actions test environment.

One other thing to note -- right now this tries native python but doesn't fail back to `tar` or the traditional tool.  I wonder if it should, or if we should have a config option like we did for the URL fetch mechanism (python vs. curl).  We've ended up having to use that one quite a bit due to python SSL not frequently being set up right -- I hope this would be more reliable.

johnwparent commented:
> One other thing to note -- right now this tries native python but doesn't fail back to `tar` or the traditional tool. I wonder if it should, or if we should have a config option like we did for the URL fetch mechanism (python vs. curl). We've ended up having to use that one quite a bit due to python SSL not frequently being set up right -- I hope this would be more reliable.

Falling all the way back to a traditional `tar` functionality would certainly be straightforward and worthwhile to implement, but personally I learn toward the config approach. Particularly with Windows it strikes me as the way to go, users could point at something like 7zip if they don't have the GNU32 project or similar on their system.

johnwparent committed:
Add decompression config opt

johnwparent committed:
Introduce and implement compression test

Refactor decomp fallback config

py2 compatibility

Fixup comp for py2

tgamblin commented on etc/spack/defaults/config.yaml:
```diff
@@ -139,6 +139,18 @@ config:
+  # Note: This can be a python module if said module is executable from a shell.
+  # ex: decompression_strategy: tar
+  # ex: decompression_strategy: 7zip
+  decompression_strategy: tar
```
This should probably be `fallback_decompression_strategy` or something like that.  The comment is confusing to me as it isn't clear what to do if I want a python module -- I thought at first that this was just a tool but the comment confused me.

johnwparent replied:
Disregard the mention of python, I had initially intended to take a different approach and forgot to remove that portion of the comment, my apologies.

tgamblin commented on lib/spack/spack/schema/config.py:
```diff
@@ -103,6 +103,12 @@
+            'decompression_strategy': {
+                'oneOf': [
+                    {'type': 'string'},
+                    {'type': 'array',
```
hmm the comment on this doesn't say it can be a list too.

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +25,189 @@ def allowed_archive(path):
+    # files with .xz style extensions due to inconsistent
+    # availability of the lzma module needed to decompress
+    # .xz files
+    if extension and not re.search(r'\.?Z', extension):
```
Why is the dot optional?  I think you want `r'.Z$'` -- currently this will find any `Z` at all in the name.

johnwparent replied:
The dot is optional as the extensions are passed in without a `.` if they are the sole extension, i.e. `.Z` is passed as `Z`, so the best solution is probably just `r'Z$'`

tgamblin commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +25,189 @@ def allowed_archive(path):
+def _system_fallback(archive_file):
+    outfile = os.path.basename(archive_file)
+    fallback = sconf.config.get('config:decompression_strategy')
+    extractor = which(fallback[0], required=True)
```
it looks like from here that the fallback can only be an executable -- so the comment and schema probably need clarifying.  I could see the fallback taking a list, though, with tar *and* 7zip, e.g.:


```yaml
config:
  fallback_decompression_tool:
    - tar
    - 7zip
```

Then you could see which one you find (and it would work on windows or unix)

johnwparent replied:
I had intended the list to be used to pass CL arguments to the utility, but I'll extend it to be both.

johnwparent committed:
Refine config blame test

Suppress vermin catch

Use lzma import as tar compression compatibility check.
This trips vermin, suppress that.

Reorder imports

Fixup import sorts

Local style checks conflicted with CI, fixup

johnwparent committed:
Improve facility of compression utility

add proper extension stripping

rm remove unpack tree

checking avoiding exp tar

rename outfile

catch improper archive extract

Catch invalid index

cleanup after compression

scheibelp commented on lib/spack/spack/test/util/spack_yaml.py:
```diff
@@ -80,7 +80,7 @@ def get_file_lines(filename):
     for line in config_blame.split("\n"):
         # currently checking only simple lines with dict keys
-        match = re.match(r"^([^:]+):(\d+)\s+([^:]+):\s+(.*)", line)
+        match = re.match(r"^([^:]+):(\d+)\s+([^#:]+):\s+(.*)", line)
```
(question) What does this change account for?

johnwparent replied:
I cannot for the life of me remember and I removed the commit message explaining why. Lets remove it and see what happens.

johnwparent replied:
This change is required to prevent the addition to the Spack yaml file from breaking the Spack yaml util tests.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+        tar = which('tar', required=True)
+        tar.add_default_arg('-oxf')
+        tar(archive_file)
+    return outfile
```
`_bunzip2` strips the extension when reporting `outfile` but this doesn't, is there a reason for that?

Also what should this return for "exploding" archives (i.e. archives that consist of multiple files not contained in a single base directory)?

johnwparent replied:
The extension stripping was at one point causing some problems with tarfiles on Windows, but after some local testing it seems like the underlying issues has been resolved.

Exploding archives are now handled. Both comments addressed in review PR.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+        return which(executor, required=True)
+    except CommandNotFoundError:
+        pass
+    return None
```
(minor) `return None` is always implied for Python functions which don't return a value, so this last statement isn't needed.

johnwparent replied:
Resolved in review commit.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+            islist = True
+        executor = try_exec(test_exec)
+        if executor:
+            args = ''
```
When specified, `args` is a list, so I think it's ideal if the default is `[]` (this means for example that you could pass `*args` into a call to `Executable`, which you can't do with `str`.

johnwparent replied:
Amended in review commit.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+    if not extractor:
+        raise CommandNotFoundError("Unable to find system fallback unpacking utility")
+    if args:
+        [extractor.add_default_arg(arg) for arg in args[1:]]
```
`args` from `derive_fallback_executable` already is computed from `strategy[1:]`, so does this actually need to be `args[1:]`? I would expect that `args` is correct and that `args[1:]` would strip off the first argument.

johnwparent replied:
I believe you are correct. Addressed in review commit.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+        fallback = [fallback]
+    extractor, args = derive_fallback_executable(fallback)
+    if not extractor:
+        raise CommandNotFoundError("Unable to find system fallback unpacking utility")
```
(minor) IMO it would slightly simplify things to just place this `raise` in `derive_fallback_executable` (if we need that function to find something, it might as well be the one to report the failure if it doesn't succeed).

johnwparent replied:
Additionally, it's not like `derive_fallback_executable` is returning anything on failure to find a program. Good point. Amended in review commit.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+    passed through to select_decompressor_for. Multiple
+    extension archives are decomposed into their component
+    extensions and the requsite decompression algorithms are
+    curried and returned as a single callable function pointer.
```
(question) is the idea that windows doesn't have any single exe that can decompress e.g. `.tar.gz`?

If that's the case, I think WRT:

```
decompressor_for(path, ext=".".join(ext_l[1:]))
```

it would be more-straightforward to create an object or function called a `two_stage_decompression` (and avoid the use of the `composer` function) since AFAIK there isn't a way to handle the case where `ext_1[1:]` has more than one element.

johnwparent replied:
The idea is that Windows _may not_ have the ability to extract a compressed archive, but will more than likely have the ability to decompress and then extract, so we piece out the process. Newer versions of Windows 10 and all of Windows 11 will not have this problem, so this can realistically be a short lived solution (I can comment to that effect if you think it's relevant) although I cannot speak to non-professional versions of Windows.  

I think I see what you're getting at but (and maybe I'm reducing the problem a little too far here) that essentially sounds like what we're doing now with a more explicit understanding that we will not handle more than two file exts (which I believe we can actually support, I just wasn't sure if the case came up in Spack).

scheibelp replied:
> I think I see what you're getting at but (and maybe I'm reducing the problem a little too far here) that essentially sounds like what we're doing now with a more explicit understanding that we will not handle more than two file exts (which I believe we can actually support, I just wasn't sure if the case came up in Spack).

My current understanding/opinions are:

* (assumption) This solution in this PR doesn't currently handle > 2 extensions (but it could easily be extended to)
* The solution is rather complex IMO in terms of handling the case of 2 extensions for Windows, so I'd prefer to see an object explicitly created for handling 2 extensions
* (assumption) I'm not aware of compression/archiving combos leading to > 2 extensions

If either of my assumptions are false, then I would revisit the value of this code. IMO it's a bit complicating to ad-hoc introduce a generic currying mechanism into the compression code though.

johnwparent replied:
My initial thought was to make it easily extensible to handle an arbitrary number of extensions but to explicitly only handle two for the time being as I also am not aware of a case with >2 extensions. You are correct that this code will ignore any more than two extensions. The ad-hoc currying is potentially pretty fragile, but I thought it would be a little easier to extend down the line (again, I can't imagine a reason for having three file extensions in this context), but I can refactor to an object no problem. Basically it boils down to how much relevancy you see the multiple extension case having (or the currying just being a little too much).

scheibelp replied:
I think this should remove the currying for now.

scheibelp commented on etc/spack/defaults/config.yaml:
```diff
@@ -139,6 +139,20 @@ config:
+  # The default fallback decompression/unpacking strategy used to
+  # unpackage archived/compressed files when current Spack strategies
+  # are unable to resolve for a particular package.
+  # Should be configured on a per system basis, as tar may not be available
```
If this should be configured on a per-system basis, then should this actually be placed in `etc/spack/defaults/windows/config.yaml` (e.g. we have similar MacOS-specific considerations placed in `etc/spack/defaults/darwin/...`)?

johnwparent replied:
It's applicable across platforms, and the compression fallback utility relies on it.

scheibelp replied:
In that case, I think the comment should be reworded (although FYI I have other comments on the need for this file so I recommend looking at the other review comments before addressing this).

scheibelp commented on etc/spack/defaults/config.yaml:
```diff
@@ -139,6 +139,20 @@ config:
+  # Should be configured on a per system basis, as tar may not be available
+  # out of the box on some systems
+  # Options are not limited to specific tools, but whatever is selected must
+  # be available on the system or on PATH on windows.
```
(minor) can the term "option" be replaced with "list element" or "item" (the former is the clearest IMO, the later is used in YAML schema syntax).

johnwparent replied:
"list element" sounds good.

scheibelp commented on lib/spack/spack/schema/config.py:
```diff
@@ -103,6 +103,20 @@
+                'oneOf': [
+                    {'type': 'string'},
+                    {'type': 'array',
+                     'items': {'oneOF': [
```
(minor) "oneOF" -> "oneOf" (I'm assuming the YAML schema accepts this but IMO more readable to have "oneOf")

johnwparent replied:
YAML will accept this, but you're right, it's cleaner to have the lowered 'f', addressed in review commit.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,174 @@
+    sys.modules = catch_modules
+
+
+@pytest.mark.xfail
```
If `scomp.decompressor_for` is supposed to raise an exception in this case, then you should use `with pytest.raises(...)`.

`xfail` is for tests that should eventually pass but we know they won't.

johnwparent replied:
Addressed in review commit.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,174 @@
+        which(util, required=True)
+    except CommandNotFoundError:
+        return False
+    return True
```
Would `which(file, required=False)` also do this? IMO that's understandable enough that this function isn't needed.

johnwparent replied:
Works for me! Cutting down the function, but leaving as a shorter alias, as IMO it makes the test noticeably cleaner.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,174 @@
+    if ext and re.match(r'bz2', ext):
+        return 'bz2' if style == "native" else "bunzip2"
+    if ext and not re.search(r'\.?Z', ext):
+        return 'tarfile' if style == "native" else "tar"
```
(I think this is only used transitively by `skip()` so if you agree with me about removing that, then this will not be worth reading)

A couple things:

* Since we're always checking if `style == native`, it should be a boolean or at least an enum (not that Python has enums but you can approximate an enum with an object and class attributes).
* What happens if none of these match? should it fail?

johnwparent replied:
`style==native` issue is resolve in review commit. If none of these match we fall back to using Tar and allowing that to fail for us, as would happen in production.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,174 @@
+@pytest.mark.parametrize('archive_file', scomp.ALLOWED_ARCHIVE_TYPES, indirect=True)
+def test_native_unpacking(tmpdir_factory, archive_file):
+    extension = scomp.extension(archive_file)
+    if not skip(extension, "native"):
```
I think it would be preferable to avoid dynamic test skipping: our CI has setup scripts which can install any needed tools, and in general we have tests that will fail without running that. It also avoids cases where those setup scripts fail or the compression algorithms are unavailable due to failures we should address.

johnwparent replied:
The idea here is to only skip if the version of Python running the test does not natively support that style of decompression/extraction (which AFAICT is limited to LZMA related material, but I wanted this to be more general), without needing to `pip install` any external modules. I think skipping here is acceptable as this unit tests checks for native python unpacking and that style may not be natively supported. I agree that the check below however, should probably be removed.

scheibelp replied:
Comprehensive coverage would require

* testing decompression using Python's built-in support
* testing decompression without it

In both cases, we'd want all those utilities to be available in the testing environment, and we should know ahead of time what decompression resources we need (in other words, what happens if one of the decompression mechanisms is skipped everywhere? In that case it wouldn't be covered).

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,174 @@
+
+
+def build_temp_archive(extension, mode, archiver, compression, archive_file, ext_dir):
+    if compression.lower() == ':z':
```
It looks like `archive_file` is prepending a `:` to get compression - why is that? Ideally when it comes to constructing and checking options IMO enums are more readable than string constructions (minor typos can lead to hard-to-debug errors when options are selected with strings).

johnwparent replied:
That's a vestigial development piece I overlooked. Removed.

scheibelp reviewed (changes requested):
I have some requests. Feel free to let me know if any of them are unclear or seem like they're headed in the wrong direction.

scheibelp commented:
@tgamblin it looks like your review has been addressed. I added some comments of my own. My intent is to dismiss your initial review once those are addressed.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,20 +30,153 @@ def allowed_archive(path):
+        if not isinstance(strategy, string_types):
+            test_exec = strategy[0]
+            islist = True
+        executor = try_exec(test_exec)
```
IMO it would be clearer to catch the `CommandNotFoundError` here, vs. implementing a function only used here that returns `None`.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,20 +30,153 @@ def allowed_archive(path):
+                else:
+                    copy(os.path.join(catch_dir, f),
+                         os.path.join(orig_dir, f))
+            shutil.rmtree(catch_dir)
```
This context manager never appears to restore the CWD.

johnwparent replied:
I believe it is restored on line 44 via `os.chdir(orig_dir)` however I could be wrong.

Edit: Apologies, I realized I hadn't pushed that change yet, the branch should be up to date now.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,20 +30,153 @@ def allowed_archive(path):
+        # catch an exploding archive
+        os.chdir(orig_dir)
+        unpacked_contents = os.listdir(catch_dir)
+        metadata = [x for x in unpacked_contents if x.startswith('.')]
```
Can this be consolidated with `URLFetchStrategy.expand`?

Minor: I'm confused about why this is being called "metadata". Some files that start with `.` are in fact useful (e..g if we archive a git repository, that includes `.git`); apparently that function also ignores this possibility so IMO as part of this you can continue to do so (a TODO comment would be useful though)

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -48,30 +188,94 @@ def _unzip(archive_file):
+    # Catch unexpected extensions and .Z files
+    # here. Python does not have native support
+    # of any kind for .Z files. In these cases, fall back
+    # to system defined strategy.
```
I've been thinking more about the basis for using a config file as a fallback and it seems somewhat unlike our general logic here: if there is a `.Z` file, then we should probably be constructing an `Executable()` from `which(7z)` in this module, vs. falling back to the config.

BTW, do any of the tests confirm the success of `7z`? I assume it would need arguments (e.g. like `tar` needs `-x` to know that it should extract).

I think that would also address another concern I have with this: the fallback config is a list of possible strategies, which right now includes tar and 7z; therefore, it seems like this might return `tar` for a `.Z` file.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -23,20 +30,153 @@ def allowed_archive(path):
+
+    with exploding_archive_catch(outfile):
+        try:
+            import tarfile
```
(question) is it possible that the decompression modules might have similar problems as `zipfile` (i.e. could this run into the same issue as https://github.com/spack/spack/pull/30407)?

johnwparent replied:
Unsure, will test for myself to determine on unix/windows.

scheibelp replied:
Methods which used to use a system utility and now prefer Python's built-in support include:

* tar
* bunzip2

Ones that still prefer a system utility:

* gunzip
* unzip

So testing the first list is most important here. Is preferring builtin support essential to this PR? Another option to reduce the complexity is to continue preferring system tools.

johnwparent replied:
gunzip should prefer Python support. Here we've just introduced backup support for the command line utility, so I'll verify that as well. Windows does not natively bundle a number of system tools needed to extract each of these extension types. This issue (particularly with bz2) is the initial motivator for this PR. We now have 7zip as a spack package, so between that and the native tar we cover most scenarios, however 7zip source code is distributed as a `.7z`, which can only be extracted by 7zip, and `.tar.xz` which Windows tar does not natively support, so we must rely on Python initially. I'm also of the opinion that for improved cross platform support we should rely more on the Python based options wherever available (Unzip is an obvious exception).

scheibelp reviewed (changes requested):
I have some high-level requests as well as some questions (it might make sense to discuss more before updating).

johnwparent committed:
Package 7zip

johnwparent committed:
tst

scheibelp commented on lib/spack/spack/schema/config.py:
```diff
@@ -90,7 +90,7 @@
                 'type': 'array',
                 'items': {'type': 'string'}
-            }
+            },
```
This looks vestigial

johnwparent committed:
Integrate 7Zip support

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -59,21 +132,129 @@ def _unzip(archive_file):
+                decompressor_for(path, ext_l[0]),
+                decompressor_for(
+                    path, ext=".".join(ext_l[1:]))
+            )
```
Is this for things like `.tar.gz`? Is there not a (single) command that can be invoked to decompress a `.tar.gz` on Windows?

johnwparent replied:
Amended

haampie commented on lib/spack/spack/util/compression.py:
```diff
@@ -119,30 +115,23 @@ def _gunzip(archive_file):
-        unzip = which(exe, required=True)
-        unzip.add_default_arg(arg)
-        unzip(archive_file)
+    exe = 'unzip'
```
Why do we even use unzip?

scheibelp replied:
A couple points:

* Some .zip archives are distributed on Linux (so that would have made sense before this PR, since AFAIK tar doesn't handle them)
* .zip files are common on Windows and zip is the most-likely compression utility that is available there (so this also makes sense for this PR)

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -308,6 +308,58 @@ def change_sed_delimiter(old_delim, new_delim, *filenames):
+    finally:
+        # catch an exploding archive
+        os.chdir(orig_dir)
+        exploding_archive_handler(tarball_container, root_path, container_path)
```
This should only run if there was no exception (that can be achieved by moving it to immediately after the `yield`)

johnwparent replied:
Good catch.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -308,6 +308,58 @@ def change_sed_delimiter(old_delim, new_delim, *filenames):
+
+
+@system_path_filter
+def exploding_archive_handler(tarball_container, root_path, source_path):
```
I think this needs a docstring explaining the intent of the arguments:

* tarball_container: where the archive was expanded to
* source_path: the intended final path for the expanded archive contents: if the archive expands to a directory, then `source_path` should include all the contents of that directory; if the archive expands to more than one file (exploding tarball), then `source_path` should contain all these files.
* root_path: the parent of source_path: if an archive expands to one non-hidden directory, then any hidden files will be moved alongside the source_path here

In that sense, it might also be possible to get away with not having a `root_path` argument (but IMO as long as the arg is documented it's fine to keep it).

johnwparent replied:
I also thought I might be able to do away with the `root_path` argument and just extrapolating from context, but was unsure if we would need it at some point (considering we're passing in components from the stage object) but if you think we can get away without it, that works for me.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -26,31 +28,102 @@ def allowed_archive(path):
+    remnant = os.path.join(os.getcwd(), archive_file)
+    lzma = [".xz", ".txz"]
+    tmp_src = os.path.join(os.getcwd(), 'tmp-src')
+    with exploding_archive_catch(os.getcwd(), tmp_src):
```
Two (somewhat contradictory) questions:

* Why don't all the decompression methods use this?
* Why do some decompression methods now handle exploding archives if the stage already does this?

johnwparent replied:
As I understand it the compression utilities that handle only decompression shouldn't need this handler as they will not explode (the gunzip and bunzip methods) but I could be wrong. All archive based methods should have it and if they don't its an oversight, I'll go back and fix that. 
To answer your second question, from what I recall, you and I had discussed this branch back in January and this change was added as a result of that conversation.

johnwparent replied:
I've removed the file based exploding archive catches.

scheibelp replied:
> As I understand it the compression utilities that handle only decompression shouldn't need this handler as they will not explode (the gunzip and bunzip methods) but I could be wrong.

After looking at those I agree with you.

> To answer your second question, from what I recall, you and I had discussed this branch back in January and this change was added as a result of that conversation.

My concern is how this works with nested `exploding_archive` context managers, I'll have to think on that more (but if the tests cover it then it should be fine).

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,171 @@
+
+ext_archive = {}  # type: Dict[str, str]
+fake_archives = [compose_to_dict('.'.join(['Foo', ext]), ext, ext_archive) for ext
+                 in scomp.ALLOWED_ARCHIVE_TYPES]
```
(minor) I think it's more straightforward to create a for-loop than to define a one-time-use function `compose_to_dict`. If you keep it, I think it would clarify if the generic method used generic terms (e.g. I have to keep reading the method to figure out what is the key and what is the value).

johnwparent replied:
Agreed, will just introduce as the for loop.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,171 @@
+                 in scomp.ALLOWED_ARCHIVE_TYPES]
+
+
+def build_temp_archive(extension, mode, archiver, compression, archive_file, ext_dir):
```
I think the tests would be more robust if you manually created the archives and placed them in a directory (vs. creating the archives as part of the test).

johnwparent replied:
Alright. I was hesitant to commit archives/compressed files to the repo, but I agree, creating them on the fly is far from ideal.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,171 @@
+
+
+@pytest.mark.parametrize('archive_file', scomp.ALLOWED_ARCHIVE_TYPES, indirect=True)
+def test_system_unpacking(tmpdir_factory, archive_file):
```
Why is this called `test_system_unpacking`? `test_unpacking` would make sense to me, but I don't get the `system` part (for example, I don't think `system` describes the decompression method, since `decompressor_for` may select a built-in-python-support mechanism for decompression).

johnwparent replied:
If you refer to the body of the test, you'll see a line that blows away the Python modules accessible to the test in the current context, forcing `decompressor_for` to fall back on system utilities. This is in contrast to the test labeled native unpacking above.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -26,31 +28,102 @@ def allowed_archive(path):
+                    os.remove(remnant)
+        except ImportError:
+            if is_windows and ext in lzma:
+                return _7zip(archive_file)
```
Returning `_7zip` inside of a function called `untar` feels off: can this be handled at the level of `select_decompressor_for`?

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -26,31 +28,102 @@ def allowed_archive(path):
 
+def _untar(archive_file):
+    """ Untar archive. Prefer native Python `tarfile`
+    but fall back to system utility failing
```
(wording) "fall back to system utility *if there is a failure to find...*"

johnwparent replied:
Amended.

scheibelp reviewed (changes requested):
Additional questions and requests. Overall

* Manually create archive files for decompression tests
* I'm concerned about coverage of decompression tests (e.g. whether they cover all Python-based vs. system-based decompression mechanisms)
* I'm somewhat confused by the handling of exploding archives both in the staging and decompression modules (combined with inconsistent handling of exploding archives in the latter)

Let me know if any of that seems to be missing the point.

johnwparent committed:
Refactor archive catch

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -59,21 +124,124 @@ def _unzip(archive_file):
+        if ext:
+            return _untar if 'xz' not in ext or not is_windows else _7zip
+
+        raise CommandNotFoundError("Cannot unpack archive with extension: %s" % ext)
```
This will only ever be raised if `bool(ext) == False`, which probably means it's an empty string (so you may want to rephrase the error message).

johnwparent replied:
Good catch.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -59,21 +124,124 @@ def _unzip(archive_file):
-    return tar
+            return _unzip
+
+        if ext and re.match(r'gz', ext):
```
Apologies can you add a note that `.tar.gz` files will not pass this check (since it's using `re.match` vs. `re.search`) - that was not clarified before this PR, but IMO it would be useful for future readers.

johnwparent replied:
Amended.

scheibelp commented on lib/spack/llnl/util/filesystem.py:
```diff
@@ -308,6 +309,62 @@ def change_sed_delimiter(old_delim, new_delim, *filenames):
+
+
+@system_path_filter
+def exploding_archive_handler(tarball_container, stage):
```
IMO this could still use a docstring (example in https://github.com/spack/spack/pull/25185#discussion_r878432736 - except now there's one fewer argument)

scheibelp reviewed (changes requested):
A few more requests (all hopefully minor)

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -59,21 +157,124 @@ def _unzip(archive_file):
+    On Unix - simply invokes select_decompressor_for
+
+    On Windows - archives with one extension are
+    passed through to select_decompressor_for. Files
```
This docstring looks out-of-date: I don't see `select_decompressor_for` anymore (nor composition of decompressors: i.e. `tar.gz` goes to `_untar`)

johnwparent replied:
Good catch.

scheibelp commented on var/spack/repos/builtin/packages/7zip/package.py:
```diff
@@ -0,0 +1,94 @@
+
+    executables = ['7z']
+
+    version('21.07', sha1='fa288b643575e55531929e7a7759b6884982556d')
```
All new package versions need to use a sha256 sum vs. a sha1 (one of the unit tests specifically checks that)

johnwparent replied:
I had noticed that failure. I'll re-checksum.

scheibelp reviewed (changes requested):
Everything looks good except a couple minor things (and https://github.com/spack/spack/pull/25185#discussion_r882219446 from prior review).

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,90 @@
+
+ext_archive = {}
+[ext_archive.update({ext: '.'.join(['Foo', ext])}) for
+    ext in scomp.ALLOWED_ARCHIVE_TYPES]
```
I think the unit tests are failing because this expects a `Foo.Z` (and we haven't added one, which I think we agreed is fine but we'll have to exclude `Foo.Z` from this list)

johnwparent replied:
I just got my VM working, I can create .Z/Tar.Z files again, so I'll upload those now.

scheibelp commented on lib/spack/spack/test/util/compression.py:
```diff
@@ -0,0 +1,90 @@
+    with working_dir(str(tmpdir)):
+        assert not os.listdir(os.getcwd())
+        util(archive_file)
+        assert os.listdir(os.getcwd())
```
What are the contents of the Foo archive? It occurred to me it would be slightly more robust to assert for example that the archive contents are what is expected (e.g. if there's a file in the archive that contains "text", read the file and check it) - that isn't essential but describing the contents would be useful IMO.

johnwparent replied:
Currently it's an empty extension less file with the name `Foo`, so I would have to regenerate all the archives with a file with contents. Not a problem, just letting you know that it may extend the life of this PR if we want to go that way. IMO we may as well while we're here.

scheibelp replied:
Ah OK apologies for not fleshing out that idea, but yes I think it would be great to place some content in the archive.

* The tar/zip ones can be a directory with a file that just has one word
* The gunzip/bunzip2/etc. ones can be a file with one word

johnwparent committed:
Add .Z example files

johnwparent committed:
archive with contents

johnwparent committed:
Add stricter test check

johnwparent committed:
.Z extensions with contents

johnwparent committed:
Format

johnwparent committed:
Handle lzma system compression on unix

johnwparent committed:
Check for lzma before native tar

johnwparent committed:
No LocalPath on py2

johnwparent committed:
Remove 'TAR' checks

johnwparent committed:
Merge branch 'support-bunzip' of https://github.com/johnwparent/spack into support-bunzip

scheibelp commented:
Thanks!

manuelakuhn commented:
This PR break the installation of some packages for me, see #31021
